### PR TITLE
Custom Environment (ready for soft review!!)

### PR DESCRIPTION
**Pull request by kottofy**

### CODE REVIEW

File: .azure-pipelines/templates/build_validation_pipeline.yml [hunk 1]
```diff
@@ -1,5 +1,53 @@
...
+  steps:
+  - script: |
+      python -m pip install --upgrade pip
+      pip install -r .azure-pipelines/requirements/build_validation_requirements.txt
+      flake8 .
+    displayName: 'Install dependencies and run linting'
+
+  - script: | 
+      echo "Building Docker container for validation with tag: ${{ parameters.image_name }}"
+      docker build -t ${{ parameters.image_name }} -f mlops/${{ parameters.model_type }}/environment/Dockerfile mlops/${{ parameters.model_type }}/environment
```
Comment (agonzalez2014) [inline .azure-pipelines/templates/build_validation_pipeline.yml:25]: per @jamesptruitt recommendation, let's be sure that either in the GH Issues / story we call out the need to research what mechanism will allow the CI agent to cache the image and avoid re-building it each time.

File: .github/workflows/build_validation_workflow.yml [hunk 2]
```diff
@@ -2,18 +2,76 @@ name: Build Validation Workflow
...
+    inputs:
+      image_name:
+        type: string
+        description: "Execution Environment"
+        required: false
+      model_type:
+        type: string
+        description: "type of model to execute"
+        required: false
+      IS_DOCKER:
```
Comment (agonzalez2014) [inline .github/workflows/build_validation_workflow.yml]: is there a reason this is ALL_CAPS when compared to the other input params?

File: .github/workflows/build_validation_workflow.yml [hunk 3]
```diff
@@ -2,18 +2,76 @@ name: Build Validation Workflow
...
+      - name: Load Python Dependencies
+        shell: bash
+        run: |
+          set -e
+          python -m pip install --upgrade pip
+          python -m pip install --upgrade -r .github/requirements/build_validation_requirements.txt      
+      - name: Lint Code
+        run: flake8
+
+      - name: Build Docker container
```
Comment (agonzalez2014) [inline .github/workflows/build_validation_workflow.yml:44]: similar comment to capture the need to explore how to cache this as a future GH issue / story

File: .github/workflows/build_validation_workflow.yml (hunk 2 above)
Comment (kottofy) [reply .github/workflows/build_validation_workflow.yml]: THERE'S NO REASON FOR IT TO BE IN ALL CAPS 😂

File: config/config.yaml [hunk 4]
```diff
@@ -94,7 +112,7 @@ deployment_configs:
       retry_timeout: 300
       deployment_base_image: mcr.microsoft.com/azureml/openmpi4.1.0-ubuntu22.04
       deployment_desc: A batch deployment
-      test_dataset_name: london_taxi_batch_test
+      test_dataset_name: nyc_taxi_batch_test
```
Comment (agonzalez2014) [inline config/config.yaml:115]: wow good catch 0,0

File: config/config.yaml [hunk 5]
```diff
@@ -109,3 +127,37 @@ deployment_configs:
...
       deployment_instance_count: 1
+
+    custom_objectdetection_batch_dev:
+      score_file_name: score.py
+      score_dir: model/custom_objectdetection/batch
+      batch_cluster_name: batchcluster
+      batch_cluster_region: eastus
+      batch_cluster_size: STANDARD_DS3_v2
+      deployment_conda_path: model/custom_objectdetection/batch_environment/conda.yml
+      endpoint_name: nyc-taxi-be-002
```
Comment (agonzalez2014) [inline config/config.yaml]: could use a different endpoint name

Comment (agonzalez2014) [reply config/config.yaml]: Same for the "endpoint_desc"

File: config/config.yaml [hunk 6]
```diff
@@ -109,3 +127,37 @@ deployment_configs:
...
+      retry_timeout: 300
+      deployment_base_image: mcr.microsoft.com/azureml/openmpi4.1.0-ubuntu22.04
+      deployment_desc: A batch deployment
+      test_dataset_name: custom_objectdetection_batch_test
+    
+    custom_objectdetection_online_dev:
+      score_file_name: score.py
+      test_file_path: model/custom_objectdetection/online_test_data/sample-request.json
+      endpoint_name: custom-objectdetection-rte-003
+      endpoint_desc: An online endpoint serving an MLflow model for online taxi
```
Comment (agonzalez2014) [inline config/config.yaml]: could use a different endpoint_desc value

Comment (agonzalez2014) [reply config/config.yaml]: same for the deployment_name, and deployment_desc (this one was empty)

File: mlops/common/get_environment.py [hunk 7]
```diff
@@ -31,7 +33,8 @@ def get_environment(
       conda_path (str): a path to a conda file with additional packages to install
       environment_name (str): a name of the environment
       description (str): a description of the environment
-
+      docker_context_path (str): The local or remote path to the the docker build context directory
```
Comment (agonzalez2014) [inline mlops/common/get_environment.py]: there are 2 "the" in the description

File: .github/workflows/custom_objectdetection_ci_pipeline.yml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+name: Custom Object Detection CI Workflow
+
+on:
+  push:
+    branches:
+      - development
+    paths:
+      - mlops/common/**
+      - mlops/**
```
Comment (agonzalez2014) [inline .github/workflows/custom_objectdetection_ci_pipeline.yml]: i might be wrong, but doesn't this line 9 effectively recursively cover all the content in the `mlops` directory? making the other lines that start with "mlops"unnecessary?

File: mlops/custom_objectdetection/src/mlops_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+    # set pipeline level compute
+    pipeline_job.settings.default_compute = cluster_name
+    pipeline_job.settings.force_rerun = True
+    # set pipeline level datastore
+    pipeline_job.settings.default_datastore = "workspaceblobstore"
+
+    return pipeline_job
+
+
+def execute_pipeline(
```
Comment (agonzalez2014) [inline mlops/custom_objectdetection/src/mlops_pipeline.py]: I know we have a story in the backlog, so just treat this comment as me expressing my excitement once it gets moved into the `mlops/common` directory 👍

File: mlops/london_taxi/src/mlops_pipeline.py [hunk 10]
```diff
@@ -284,9 +284,9 @@ def prepare_and_execute(
         config.aml_config["subscription_id"],
         config.aml_config["resource_group_name"],
         config.aml_config["workspace_name"],
-        config.environment_configuration["env_base_image"],
-        pipeline_config["conda_path"],
-        pipeline_config["aml_env_name"],
+        env_base_image_name=config.environment_configuration["env_base_image"],
```
Comment (agonzalez2014) [inline mlops/london_taxi/src/mlops_pipeline.py]: i have no objections, but the dstoolkit-v2 team may ask why the change in syntax.

**Review by agonzalez2014 — APPROVED** [review]
I would encourage to think about what changes to the README would inform the consumers on what you are introducing. 

Great work! Yes, there is a lot of duplicate code, but I think this will be a worthwhile discussion with the dstoolkit-v2 team. Because there may be the opportunity in the future to surface the need for `src/commons` directory, as well as even potentially a `model/default` directory to explore more code reuse across model src code and deployment/hosting

File: config/config.yaml (hunk 4 above)
Comment (kottofy) [reply config/config.yaml:115]: I do what I can :)

File: config/config.yaml (hunk 6 above)
Comment (kottofy) [reply config/config.yaml]: updated

File: mlops/common/get_environment.py (hunk 7 above)
Comment (kottofy) [reply mlops/common/get_environment.py]: updated

File: mlops/london_taxi/src/mlops_pipeline.py (hunk 10 above)
Comment (kottofy) [reply mlops/london_taxi/src/mlops_pipeline.py]: I named it back.. no need to change it. good catch

File: config/config.yaml (hunk 5 above)
Comment (kottofy) [reply config/config.yaml]: updated

File: .github/workflows/custom_objectdetection_ci_pipeline.yml (hunk 8 above)
Comment (kottofy) [reply .github/workflows/custom_objectdetection_ci_pipeline.yml]: yes.. I updated all of the pipelines. good catch!
